kmod/core/core.c

```diff
@@ -583,29 +585,57 @@ static int kpatch_kallsyms_callback(void *data, const char *name,
 {
 	struct kpatch_kallsyms_args *args = data;
 
-	if (args->addr == addr && !strcmp(args->name, name))
+	if (strcmp(args->name, name))
+		return 0;
+
+	args->addr = addr;
+	args->count++;
+
+	/*
+	 * Finish the search when the symbol is found for the desired position
+	 * or the position is not defined for a non-unique symbol.
+	 */
+	if ((args->pos && (args->count == args->pos)) ||
+	    (!args->pos && (args->count > 1))) {
 		return 1;
+	}
 
 	return 0;
 }
 
-static int kpatch_verify_symbol_match(const char *name, unsigned long addr)
+static int kpatch_verify_symbol_match(const char *name,
```

Comment:
This should probably be made 100% identical to klp_find_object_symbol(), with its objname parameter, and renamed accordingly.  And the same for making kpatch_kallsyms_callback() similar to klp_find_callback().  Then I think we can get rid of kpatch_find_module_symbol() and symbol resolution will be more nicely klp-esque.